tests/templates/test_subroutines/test_adder.py

```diff
@@ -0,0 +1,167 @@
+        [
+            (
+                6,
+                [0, 1, 2],
+                8,
+                [4],
+            ),
+            (
+                1,
+                [0, 1, 2, 3],
+                9,
+                [4],
+            ),
+            (
+                2,
+                [0, 1, 4],
+                4,
+                [3],
+            ),
+            (
+                -3,
+                [0, 1, 4],
+                4,
+                [3],
+            ),
+            (
+                10,
+                [0, 1, 2, 5],
+                9,
+                [3],
+            ),
+            (
+                1,
+                [0, 1, 2],
+                7,
+                [3],
+            ),
+            (
+                6,
+                [0, 1, 2, 3],
+                None,
+                [4],
+            ),
+        ],
+    )
+    def test_operation_result(
+        self, k, x_wires, mod, work_wire
+    ):  # pylint: disable=too-many-arguments
+        """Test the correctness of the PhaseAdder template output."""
+        dev = qml.device("default.qubit", shots=1)
+        if mod is None:
+            max = 2 ** len(x_wires)
+        else:
+            max = mod
+            extra_wire=[100]
+            x_wires=x_wires+extra_wire
+            print(x_wires)
+        
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.BasisEmbedding(x, wires=x_wires)
+            Adder(k, x_wires, mod, work_wire)
+            return qml.sample(wires=x_wires)
+        for x in range(0, max):
+            assert np.allclose(
+                sum(bit * (2**i) for i, bit in enumerate(reversed(circuit(x)))), (x + k) % max
+            )
+
+    @pytest.mark.parametrize(
+        ("k", "x_wires", "mod", "work_wire", "msg_match"),
+        [
+            (
+                3,
+                [0, 1, 2, 3, 4],
+                12,
+                [4],
+                "work_wire should not be included in x_wires.",
+            ),
+        ],
+    )
+    def test_wires_error(
+        self, k, x_wires, mod, work_wire, msg_match
+    ):  # pylint: disable=too-many-arguments
+        """Test an error is raised when some wire in work_wire is in wires"""
+        with pytest.raises(ValueError, match=msg_match):
+            Adder(k, x_wires, mod, work_wire)
+
+    def test_decomposition(self):
+        """Test that compute_decomposition and decomposition work as expected."""
+
+        k=2
+        mod=7
+        x_wires=[0,1,2]
+        work_wire=[3]
+        Adder_decomposition = Adder(k, x_wires, mod, work_wire).compute_decomposition(
+            k, x_wires, mod, work_wire
+        )
+        op_list = []
+        op_list.append(qml.QFT(x_wires))
+        op_list.append(PhaseAdder(k, x_wires, mod, work_wire))
+        op_list.append(qml.adjoint(qml.QFT)(x_wires))
+
+        for op1, op2 in zip(Adder_decomposition, op_list):
+            qml.assert_equal(op1, op2)
+
```

Comment:
Do we need grad tests?

Reply:
I don't think so, since grad_method=None